alphazero/mcts/mcts.py

```diff
@@ -0,0 +1,282 @@
+        player_layer = torch.full(board.size(), player)
+        return torch.stack((player_1, player_2, player))
+
+    ## TODO: Flesh this out and move
+
+
+class MCTSNode:
+    """
+    Node Class for MCTS tree search
+    """
+
+    def __init__(
+        self,
+        state: torch.Tensor,
+        prior: float = 0,
+        action: torch.Tensor = torch.Tensor([]),
+    ):
+
+        self._action = action
+        self._state = state
+        self._children: List[MCTSNode] = []
+        self._value = None
+        self._policy = None
+        self.n_visit = 0
+        self.total_value = 0
+        self.is_terminal = False
+        self.prior = prior
+        self.parent = None
+
+    def add_child(self, child: MCTSNode):
+        """
+        Adds a child to node instance
+
+        Args:
+            child (MCTSNode): Node to be added as child
+        """
+        self._children.append(child)
+        child.parent = self
+
+    def rollout(self, game: GameClassType) -> int:
+        """
+        Performs a rollout from self, selecting random actions from a uniform prior until game ends
+        Args:
+            game (GameClassType): Game that is currently played
+
+        Returns:
+            result (int): Final result of rollout
+        """
+        state = self._state
+        while not game.is_game_over(state):
+            legal_actions = game.get_legal_moves(state)
+            action = sample_tensor_indices(legal_actions, 1)[0]
+            player = 1 if (state == 0).sum() > (state == 1).sum() else 0
+            state = game.board_after_move(state, player, action)
+        result = game.result(state)
+        return result
+
+    def calc_policy_value(
+        self, rollout: bool, game: GameClassType, network: torch.nn.Module = None
+    ) -> (torch.Tensor, float):
+        """
+        Calculates the policy and value for the node. If rollout performs a rollout, else uses network to get policy and value
+        Args:
+            rollout (bool): Whether to perform a rollout or not
+            game (GameClassType): Game that is played
+            network (torch.nn.Module, optional): Optional, must be set if rollout is false
+
+        Returns:
+            policy (torch.Tensor): The calculated policy
+            value (float): The calculated value
+        """
+        if not rollout:
+            # TODO: Normalize probabilities after filter legal actions
+            self._policy, self._value = network(self._state)
+        else:
+            self._value = self.rollout(game)
+            policy = game.get_legal_moves(self._state)
+            self._policy = policy.float() / policy.sum()
+        return self._policy, self._value
+
+    def expand(self, game: GameClassType):
+        """
+        Expands node
+        Args:
+            game (GameClassType): Game that is played
+        """
+        legal_actions = game.get_legal_moves(self._state)
+        for index in legal_actions.nonzero():
```

Comment:
are we sure this works with False? I have no idea what goes on internally in pytorch so I don't want to assume.

Reply:
It seems to work with byte tensors, so think so?

Reply:
we should probably have a test for this just in case then

Reply:
![image](https://user-images.githubusercontent.com/18634426/84978054-94e71300-b0f1-11ea-929f-b98b9c3f5bd8.png)